simapp/abci.go

```diff
@@ -2,53 +2,161 @@ package simapp
+	Vote_SKIP
+)
+
+func NewVoteExtensionHandler(app *SimApp) *VoteExtensionHandler {
+	return &VoteExtensionHandler{app: app}
 }
 
-func (h *VoteExtensionHandler) SetHandlers(bApp *baseapp.BaseApp) {
-	bApp.SetExtendVoteHandler(h.ExtendVote())
-	bApp.SetVerifyVoteExtensionHandler(h.VerifyVoteExtension())
+func (v VoteExtensionHandler) SetVoteExtHandler(modVoteExtHandler map[string]func(ctx sdk.Context, req *abci.RequestExtendVote) (*abci.ResponseExtendVote, error)) {
+	v.modVoteExtHandler = modVoteExtHandler
 }
 
-func (h *VoteExtensionHandler) ExtendVote() sdk.ExtendVoteHandler {
-	return func(_ sdk.Context, req *abci.RequestExtendVote) (*abci.ResponseExtendVote, error) {
-		buf := make([]byte, 1024)
+func (v VoteExtensionHandler) SetVerifyVoteExtHandler(modVerifyVoteExtHandler map[string]func(ctx sdk.Context, req *abci.RequestVerifyVoteExtension) (*abci.ResponseVerifyVoteExtension, error)) {
+	v.modVerifyVoteExtHandler = modVerifyVoteExtHandler
+}
 
-		_, err := rand.Read(buf)
+// PrepareProposal -> Verify Prev. Block's VEs (check for 2/3+ majority ) -> Encode VEs in tx bytes
+func (app *SimApp) NewPrepareProposalHandler() sdk.PrepareProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestPrepareProposal) (*abci.ResponsePrepareProposal, error) {
+
+		// Log the event record stored in previous height
+		event, err := app.StakingKeeper.GetEventRecord(ctx)
 		if err != nil {
-			return nil, fmt.Errorf("failed to generate random vote extension data: %w", err)
+			fmt.Println("Error occurred while fetching Event Record from DB: ", err)
+		}
+		fmt.Println("Event stored in height : ", ctx.BlockHeight()-1, "event ID: ", event.Id, "event block no.: ", event.BlockNumber)
+
+		// Validate VE sigs and check whether they have 2/3+ majority
+		hasTwoThirdsSigs := true
+		if err := app.StakingKeeper.ValidateVoteExtensions(ctx, ctx.BlockHeight(), ctx.ChainID(), req.LocalLastCommit.Votes, req.LocalLastCommit.Round); err != nil {
+			fmt.Println("PrepareProposal: Error occurred while validating VEs: ", err)
+			hasTwoThirdsSigs = false
 		}
 
+		var txs [][]byte
+		rawTxs := req.Txs
+		// Only include VE if there's majority
+		if hasTwoThirdsSigs {
+			fmt.Println("PrepareProposal: INJECTING VEs!")
+			// VoteExtensions will be encoded before other txs
+			extVoteInfo, err := json.Marshal(req.LocalLastCommit.Votes)
+			if err != nil {
+				return nil, err
+			}
+			txs = append(txs, extVoteInfo)
+		}
+
+		// encode the txs
+		txs = append(txs, rawTxs...)
+		return &abci.ResponsePrepareProposal{Txs: txs}, nil
+
+	}
+
+}
+
+// Verify whether the proposal has VEs since a malicious proposer could censor votes.
+// Check for 2/3+ VE sigs and reject the proposal in case we don't have a majority.
+// A downside is liveness can get negatively affected.
+func (app *SimApp) NewProcessProposalHandler() sdk.ProcessProposalHandler {
+	return func(ctx sdk.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
+		if req.Height > 5 {
```

Comment:
As in the the `staking/keeper.go` helper, use
```go
    cp := ctx.ConsensusParams()
	extsEnabled := cp.Abci != nil && currentHeight >= cp.Abci.VoteExtensionsEnableHeight && cp.Abci.VoteExtensionsEnableHeight != 0
``````

Reply:
@VAIBHAVJINDAL3012